src/timezones.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <port.h>
+#include <catalog/pg_type.h>
+#include <executor/spi.h>
+#include <utils/builtins.h>
+
+#include "timezones.h"
+
+static SPIPlanPtr savedPlanNormalizeTimezone = NULL;
+
+/*
+ * For the given TZ name or abbreviation return the normalized TZ name.
+ * For invalid TZ name NULL is returned.
+ * Examples:
+ * - normalized_timezone_name("Europe/Moscow") = "Europe/Moscow"
+ * - normalized_timezone_name("MSK") = "Europe/Moscow"
+ * - normalized_timezone_name("Europe/Ololondon") = NULL
+ */
+const char *
+ts_validate_and_normalize_timezone_name(const char *tz_name)
+{
+	int res;
+	bool isnull;
+	char *buffer;
+	Datum qvalues[] = { CStringGetTextDatum(tz_name) };
+
+	if (SPI_connect() != SPI_OK_CONNECT)
+		elog(ERROR, "could not connect to SPI");
+
+	if (savedPlanNormalizeTimezone == NULL)
+	{
+		Oid argtypes[] = { TEXTOID };
+		savedPlanNormalizeTimezone = SPI_prepare("SELECT name FROM pg_catalog.pg_timezone_names "
```

Comment:
Can we do this with an internal catalog scan like we normally do for native code or use a utility function?

If you look at the view `pg_timezone_names` it is actually just generated from calling the same named function, which in turn enumerates the timezones with `pg_tzenumerate_start()`, which you could call directly in native code in a similar way.

There's also a PostgreSQL function called `FetchDynamicTimeZone` in `datetime.c`  that shows how you can use `pg_tzset` to lookup timezones. It is using an internal cache.

Reply:
Also, I doubt that doing a manual scan of the table is necessarily better than executing a cached SPI plan.

Reply:
@afiskon I think the main @erimatnor's concern is because as far I understood since I joined the team is we try to avoid use SPI as far we can.

Reply:
@fabriziomello because.... ?

Reply:
I don't see the benefit of SPI here either. It's not faster, the code stack is deeper, more things can go wrong and it's not cleaner to look at as far as code maintainability/readability is concerned.

Reply:
`pg_tzenumerate_start()` is basically the thing that gets called underneath the hood in the SPI query, so might as well just call that directly for cleaner code.

I am also wondering, however, if there isn't a way to normalize time zones with existing built-in API functions? The query here essentially reads all the timezone files from disk, which seems a bit inefficient. Isn't there a place where this information is cached?

Reply:
@erimatnor there are less than 600 rows in the table and the query takes about 60 ms because the table is cached in memory. It reads nothing from the disk.

@mfundul This is an odd argument considering the fact that the benefits of the alternative approach were not stated so far either.

> I am also wondering, however, if there isn't a way to normalize time zones with existing built-in API functions?

That's in fact a good question. I will investigate.

Reply:
I see benefits in not calling and not error checking the following functions:
`SPI_connect()`, `SPI_prepare()`, `SPI_keepplan()`, `SPI_getbinval()`, `SPI_palloc()`, `SPI_finish()` or using OIDs such as `argtypes`.
As far as readability is concerned, it doesn't look good. Something like the following would read better in my opinion:
```
	for (tzenum = pg_tzenumerate_start(), tz = pg_tzenumerate_next(tzenum) ; tz; tz = pg_tzenumerate_next(tzenum))
	{
		if (timestamp2tm(GetCurrentTransactionStartTimestamp(),
						 &tzoff, &tm, &fsec, &tzn, tz) != 0)
			continue;

        name = pg_get_timezone_name(tz);
        abbrev = tzn ? tzn : "";

		if (!strcmp(tz_name, name) || !strcmp(tz_name, abbrev))
		{
			found = true;
			tz_name = CStringGetTextDatum(name);
			break;
		}
	}
```

Reply:
Thanks for all the great feedback! Here is a new version of the patch based on pg_tzenumerate_* as @mfundul suggested.